README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ggdal

<!-- badges: start -->

[![R-CMD-check](https://github.com/hypertidy/ggdal/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/hypertidy/ggdal/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

The goal of ggdal is to get background image data for ggplot2.

No futzing around with map tiles or zooms or other programs.

## Installation

You can install the development version of ggdal from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("hypertidy/ggdal")
```

# Example

Use ‘dsn = “osm”’ or “virtualearth”, or input custom data source, this
is a source readable by GDAL. Some examples are found in the sds package
available on [hypertidy/sds](https://github.com/hypertidy/sds.git).

Use any map projection. Currently examples only provided for use with sf
objects, WIP.

``` r
library(ggplot2)
library(ggdal)
data(iw)

ggplot() +
  annotation_gdal(dsn = "osm") +
  geom_sf(data = sf::st_transform(iw, "EPSG:3577"), fill = NA, col = "grey50")
```

<img src="man/figures/README-example-1.png" width="100%" />

``` r


ggplot() +
  annotation_gdal(dsn = "virtualearth") +
  geom_sf(data = sfdct::antarctica, fill = NA, col = "grey50") + coord_sf(xlim = c(-6378137, 6378137), 
                                                                          ylim = c(-6378137, 6378137))
#> old-style crs object detected; please recreate object with a recent sf::st_crs()
#> old-style crs object detected; please recreate object with a recent sf::st_crs()
```

<img src="man/figures/README-example-2.png" width="100%" />

## fun

For a bit of a lark, get your favourite dataset ‘x’ and run this code a
few times.

``` r
laea <- function(x) {
  bb <- sf::st_bbox(x)
  sf::st_transform(x, sprintf("+proj=laea +lon_0=%f lat_0=%f", mean(bb[c(1, 3)]),   mean(bb[c(2, 3)])))
}
x <- sf::read_sf("https://datahub.io/core/geo-countries/r/countries.geojson")
srcs <- c(sds::wms_openstreetmap_tms(), 
          sds::wms_virtualearth_street(), 
          sds::wms_arcgis_mapserver_ESRI.WorldImagery_tms())

for (i in 1:5) {
  i <- i + 1
plt <- ggplot() + 
  annotation_gdal(sample(srcs, 1)) + 
  geom_sf(data = laea(sf::st_convex_hull(dplyr::sample_n(x, 1))), fill = NA, colour = "hotpink", linewidth = 2)
print(plt)



}
```

<img src="man/figures/README-fun-1.png" width="100%" /><img src="man/figures/README-fun-2.png" width="100%" /><img src="man/figures/README-fun-3.png" width="100%" /><img src="man/figures/README-fun-4.png" width="100%" /><img src="man/figures/README-fun-5.png" width="100%" />

## Code of Conduct

Please note that the ggdal project is released with a [Contributor Code
of
Conduct](https://contributor-covenant.org/version/2/1/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.


NEWS.md
# ggdal 0.0.3

* Remove whatarelief in favour of vapour. 

* Now importing hypertidy/sds package rather than using built-in sources. 


# ggdal 0.0.2

* Removed old deprecated stuff. 

# ggdal 0.0.1

* New function 'annotation_gdal()' adapted from `ggspatial::annotation_map_tile()` to add a simple background image
via GDAL warper API. Built in OSM streetmap and Virtual earth satellite options, or use own 'dsn'. 

All prior use of 'geom_from_gdal' is deprecated. 



tryout.R
local_image_path <- system.file("r_logo.png", package = "ggpath")

img <- gdal_get(local_image_path)


vp <- grid::viewport(
        x = grid::unit(0.5, "npc"),
        y = grid::unit(0.5,"npc"),
        width = grid::unit(0.2, "npc"),
        height = grid::unit(0.3, "npc"),
        angle = 0
      )
grid.draw(rasterGrob(img, vp = vp))




# compute path of an R logo file shipped with ggpath
local_image_path <- system.file("r_logo.png", package = "ggpath")

# create dataframe with x-y-coordinates and the above local path
plot_data <- data.frame(x = 0, y = c(-1, 1), gdal = local_image_path, px = 26, py = 18, resample = c("near", "cubic"))

# plot images directly from local path
ggplot(plot_data, aes(x = x, y = y, px = px, py = py, resample = resample)) +
  geom_from_gdal(aes(gdal = gdal), width = 0.3, height = .3) +
  coord_cartesian(xlim = c(-1, 1), ylim = c(-2, 2)) +
  theme_minimal()


R/ggdal.R


#' @importFrom vapour vapour_raster_info vapour_warp_raster vapour_vrt
#' @importFrom palr d_pal
#' @rdname ggdal-deprecated
#' @importFrom grDevices dev.size hcl.colors
gdal_get <- function(x, dm = NULL, ..., extent = NULL, projection = NULL) {
  info <- vapour::vapour_raster_info(x)
  if (diff(info$extent)[3] < 0) {
    x <- vapour::vapour_vrt(x, extent = info$extent[c(1, 2, 4, 3)])
  }
  info <- vapour::vapour_raster_info(x)
  if (is.null(dm)) dm <- dev.size("px")
  data <- vapour::vapour_warp_raster(x, extent = info$extent, dimension = dm, bands = 1, ...)[[1L]]
  data <- palr::image_pal(as.numeric(data), col = hcl.colors(56))
  structure(list(data = matrix(data, dm[2L], byrow = TRUE),
       info = info), class = c("gdal-data", "gdal", "list"))
}
#' @rdname ggdal-deprecated
#' @export
print.gdal <- function(x, ...) {
  print(x$info$dimension)
  print(x$info$extent)
  print(sprintf("hasdata: %s", as.character(!is.null(x$data))))
}
#' @rdname ggdal-deprecated
#' @export
#' @importFrom grDevices as.raster
as.raster.gdal <- function(x, ...) {
  .Deprecated("annotation_gdal")
  out <- x$data
  if (!is.character(out)) {
    out <- array(palr::image_pal(x$data, ...), dim(x$data)[1L])
  }
  #ximage(out)
  as.raster(out)
}


# INTERNAL HELPER THAT BUILDS THE GROBS FOR GeomFromGDAL - ripped off from ggpath
#' @importFrom grid rasterGrob viewport unit
#' @importFrom rlang caller_env
build_gdal_grobs <- function(i, alpha, colour, gdal, px, py, resample, data,
                        is_theme_element = FALSE,
                        call = rlang::caller_env()) {

  img <- gdal_get(gdal[i], dm = c(px[i], py[i]), resample = resample[i])

  # theme elements require justification outside the viewport
  # so we have to do this twice here
  if(isFALSE(is_theme_element)){
    grid::rasterGrob(
      img,
      vp = grid::viewport(
        x = grid::unit(data$x[i], "native"),
        y = grid::unit(data$y[i], "native"),
        width = grid::unit(data$width[i], "npc"),
        height = grid::unit(data$height[i], "npc"),
        just = c(data$hjust[i], data$vjust[i]),

        angle = data$angle[i]
      ),
      interpolate = FALSE,
      name = paste0("ggdal.grob.", i)
    )
  } else if (isTRUE(is_theme_element)){
    grid::rasterGrob(
      img,
      vp = grid::viewport(
        x = grid::unit(data$x[i], "npc"),
        y = grid::unit(data$y[i], "npc"),
        width = grid::unit(data$width[i], "npc"),
        height = grid::unit(data$height[i], "npc"),
        angle = data$angle[i]
      ),
      just = c(data$hjust[i], data$vjust[i]),
      interpolate = FALSE,
      name = paste0("ggdal.grob.", i)
    )
  }
}



#' ggplot2 Layer for Visualizing Images from GDAL sources (file path, URLs, database connection strings, VRT text in memory, etc. )
#'
#' @description This geom is used to plot raster data from a GDAL source
#'   of points in a ggplot. Currently requeires x,y aesthetics as well as a gdal DSN, but generally these can
#' obtained from the source, and a graphics device can dictate that the GDAL warper provides the data in a given
#' extent ....
#'
#' @inheritParams ggplot2::geom_point
#' @section Aesthetics:
#' `geom_from_gdal()` understands the following aesthetics (required aesthetics are in bold):
#'  FIXME: px, py, resample, etc - do we want to have this graphics placement or totally override with xmin, xmax, ymin, ymax, etc
#' \itemize{
#'   \item{**x**}{ - The x-coordinate.}
#'   \item{**y**}{ - The y-coordinate.}
#'   \item{**path**}{ - a file path, url, raster object or bitmap array.}
#'   \item{`alpha = NULL`}{ - The alpha channel, i.e. transparency level, as a numerical value between 0 and 1.}
#'   \item{`colour = NULL`}{ - The image will be colorized with this colour. Use the special character `"b/w"` to set it to black and white. For more information on valid colour names in ggplot2 see <https://ggplot2.tidyverse.org/articles/ggplot2-specs.html?q=colour#colour-and-fill>}
#'   \item{`angle = 0`}{ - The angle of the image as a numerical value between 0° and 360°.}
#'   \item{`hjust = 0.5`}{ - The horizontal adjustment relative to the given x coordinate. Must be a numerical value between 0 and 1.}
#'   \item{`vjust = 0.5`}{ - The vertical adjustment relative to the given y coordinate. Must be a numerical value between 0 and 1.}
#'   \item{`width = 1.0`}{ - The desired width of the image in `npc` (Normalised Parent Coordinates).
#'                           The default value is set to 1.0 which is *big* but it is necessary
#'                           because all used values are computed relative to the default.
#'                           A typical size is `width = 0.1` (see below examples).}
#'   \item{`height = 1.0`}{ - The desired height of the image in `npc` (Normalised Parent Coordinates).
#'                            The default value is set to 1.0 which is *big* but it is necessary
#'                            because all used values are computed relative to the default.
#'                            A typical size is `height = 0.1` (see below examples).}
#' }
#' @param ... Other arguments passed on to [ggplot2::layer()]. These are
#'   often aesthetics, used to set an aesthetic to a fixed value. See the below
#'   section "Aesthetics" for a full list of possible arguments.
#' @return A ggplot2 layer ([ggplot2::layer()]) that can be added to a plot
#'   created with [ggplot2::ggplot()].
#' @rdname ggdal-deprecated
#' @export
geom_from_gdal <- function(mapping = NULL, data = NULL,
                           stat = "identity", position = "identity",
                           ...,
                           na.rm = FALSE,
                           show.legend = FALSE,
                           inherit.aes = TRUE) {

  .Deprecated("annotation_gdal")
  ggplot2::layer(
    data = data,
    mapping = mapping,
    stat = stat,
    geom = GeomFromGDAL,
    position = position,
    show.legend = show.legend,
    inherit.aes = inherit.aes,
    params = list(
      na.rm = na.rm,
      ...
    )
  )
}

#' @name ggdal-package
#' @export
GeomFromGDAL <- ggplot2::ggproto(
  "GeomFromGDAL", ggplot2::Geom,
  required_aes = c("x", "y", "gdal"),
  # non_missing_aes = c(""),
  default_aes = ggplot2::aes(
    alpha = NULL, colour = NULL, angle = 0, hjust = 0.5,
    vjust = 0.5, width = 1.0, height = 1.0, px = 64, py = 64, resample = "bilinear"
  ),
  draw_panel = function(data, panel_params, coord, na.rm = FALSE) {
    data <- coord$transform(data, panel_params)

    grobs <- lapply(
      seq_along(data$gdal),
      build_gdal_grobs,
      alpha = data$alpha,
      colour = data$colour,
      gdal = data$gdal,
      px = data$px,
      py  = data$py,
      resample = data$resample,
      data = data
      )

    class(grobs) <- "gList"

    grid::gTree(children = grobs)
  },
  draw_key = function(...) grid::nullGrob()
)


R/annotation_gdal.R
.ideal_dim <- function(dimension = dev.size("px"), extent = c(0, 1, 0, 1)) {
  rat <- diff(extent[1:2])/diff(extent[3:4])
  #print(paste("rat:", rat))
  if (rat < 1) {
    width <- dimension[2L] * rat
    height <- dimension[2L]
  }
  if (rat >= 1) {
    width <- dimension[1L]
    height <- dimension[1L]/rat
  }
  #dm <- as.integer(c(width, height))
  #print(dm)
  #print(diff(par("plt"))[c(1, 3)] * dm0)

  ## using plt is removing too much
  #as.integer(round(diff(par("plt"))[c(1, 3)] * c(width, height)))
  as.integer(c(width, height) * .95)
  #c(dimension[1L], 0)
}



#' Add background imagery
#'
#' Uses OpenStreetMap or VirtualEarth to add background imagery, or a custom source via 'dsn'.
#'  If you are publishing
#' a map using these tiles, make sure to use the proper attribution
#' (e.g., "Copyright OpenStreetMap contributors" when using an
#' OpenStreetMap-based tile set).  Ditto for VirtualEarth or any dsn you use.
#'
#' @param dsn The map source (currently 'osm' or 'virtualearth' are built-in -streetmap, or imagery, otherwise use a GDAL DSN - see package {sds} for some helpers)
#' @param interpolate Passed to [grid::rasterGrob()]
#' @param alpha Use to make this layer semi-transparent
#' @param resample resample algorithm for the GDAL warper
#' @param data,mapping Specify data and mapping to use this geom with facets
#'
#' @return A ggplot2 layer
#' @export
#' @importFrom vapour gdal_raster_image
#' @importFrom scales alpha
#' @importFrom grDevices as.raster rgb col2rgb
#' @importFrom grid rasterGrob
#' @examples
#' \donttest{
#' library(ggplot2)
#' data(iw)
#'
#' ggplot() +
#'   annotation_gdal() +
#'   geom_sf(data = sf::st_transform(iw, "EPSG:3857"), fill = NA, col = "grey50")
#'
#'  ggplot() +
#'   annotation_gdal("virtualearth") +
#'   geom_sf(data = iw, fill = NA, col = "grey50")
#'
#'
#' pts <- do.call(cbind, maps::map(plot = FALSE)[1:2])
#' pts <- pts[!is.na(pts[,1]), ]
#' pts <- pts[seq(1, nrow(pts), length.out = 8000), ]
#' sf <- sf::st_sf(geom = sf::st_sfc(sf::st_multipoint(pts), crs = "OGC:CRS84"))
#' ggplot() +
#'   annotation_gdal(dsn = "virtualearth") +
#'   geom_sf(data = sf::st_transform(sf, "+proj=laea +lon_0=147 +lat_0=-42"),
#'   fill = NA, col = "yellow", pch = ".")
#'
#' pts2 <- pts[pts[,1] > 50 & pts[,1] < 120 & pts[,2] < -20 & pts[,2] > -45, ]
#' sf <- sf::st_sf(geom = sf::st_sfc(sf::st_multipoint(pts2), crs = "OGC:CRS84"))
#' ggplot() +
#'   annotation_gdal(dsn = "osm") +
#'   geom_sf(data = sf::st_transform(sf, "EPSG:3577"), fill = NA, col = "yellow", pch = ".")
#' pts3 <- pts[ pts[,2] < -20, ]
#' sf <- sf::st_sf(geom = sf::st_sfc(sf::st_multipoint(pts3), crs = "OGC:CRS84"))
#' wms_arcgis_mapserver_tms <- arcgis_mapserver_imgery()
#' ggplot() +
#'   annotation_gdal(dsn = wms_arcgis_mapserver_tms, resample ="lanczos") +
#'   geom_sf(data = sf::st_transform(sf, "EPSG:3031"), fill = NA, col = "hotpink", pch = 19, cex = 0.2)
#'
#' }
#'
annotation_gdal <- function(dsn = c("osm", "virtualearth"), resample = "bilinear",
                            interpolate = FALSE, data = NULL, mapping = NULL, alpha = 1) {

  if(is.null(data)) {
    data <- data.frame(dsn = dsn, resample = resample, stringsAsFactors = FALSE)
    mapping <- ggplot2::aes(dsn = dsn, resample = resample)
  }

  c(
    ggplot2::layer(
      data = data,
      mapping = mapping,
      geom = GeomGdal,
      stat = "identity",
      position = "identity",
      params = list(
        interpolate = interpolate,
        alpha = alpha
      ),
      inherit.aes = FALSE,
      show.legend = FALSE
    )
  )
}

#' @export
#' @rdname annotation_gdal
GeomGdal <- ggplot2::ggproto(
  "GeomGdal",
  ggplot2::Geom,

  extra_params = "",

  handle_na = function(data, params) {
    data
  },

  default_aes = ggplot2::aes(dsn = "osm", resample = "bilinear" ),

  draw_panel = function(
    data, panel_params, coordinates,  interpolate = TRUE, alpha = 1
  ) {


    coord_crs <- sf::st_crs(panel_params$crs)
    ex <- c(panel_params$x_range, panel_params$y_range)
    prj <- coord_crs$wkt


    ## here we need the dims to be smart, just pivot on the aspect ratio to a constant size for now
    dm <- .ideal_dim(extent = ex)
    src <- NULL

    dsn <- as.character(data$dsn[[1L]])
    resample <- as.character(data$resample[[1L]])
    if (dsn == "osm") {
      src <- sds::wms_openstreetmap_tms()
    }
    if (dsn == "virtualearth") {
      src <- sds::wms_virtualearth()
    }

    ## the user has passed in a DSN

    if (is.null(src))  src <- dsn
    ## dm might have a zero in it, so we use the record attribute "dimension"
    #print(paste("nr/nc:", dm[1]/dm[2]))
    #print(paste("dm", dm))

    #print(paste("px:", dev.size("px")))
    dpix <- vapour::gdal_raster_image(src, target_ext  = ex, target_crs = prj, target_dim = dm, resample = resample)
    img <- as.raster(matrix(dpix[[1L]], attr(dpix, "dimension")[2L], byrow = TRUE))
    if (alpha < 1) {
      img <- as.raster(matrix(scales::alpha(img, alpha = alpha), dm[2], byrow = TRUE))
    }
    ## again, override the extent (so we can allow no extent to be passed in )
    ex <- attr(dpix, "extent")
    corners <- data.frame(x = ex[1:2], y = ex[3:4])

    # transform corners to viewport cs
    corners_trans <- coordinates$transform(corners, panel_params)
    x_rng <- range(corners_trans$x, na.rm = TRUE)
    y_rng <- range(corners_trans$y, na.rm = TRUE)

    # return raster grob of the img
    grid::rasterGrob(
      img,
      x_rng[1], y_rng[1],
      diff(x_rng), diff(y_rng), default.units = "native",
      just = c("left","bottom"), interpolate = interpolate
    )
  }
)

